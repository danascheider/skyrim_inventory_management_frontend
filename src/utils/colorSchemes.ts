export interface ColorScheme {
  schemeColorDarkest: string
  schemeColorDark: string
  schemeColorLight: string
  schemeColorLightest: string
  hoverColorDark: string
  hoverColorLight: string
  hoverColorLightest: string
  textColorPrimary: string
  textColorSecondary: string
  textColorTertiary: string
  borderColor: string
}

export const YELLOW: ColorScheme = {
  schemeColorDarkest: '#ffbf00',
  schemeColorDark: '#ffcb32',
  schemeColorLight: '#ffd866',
  schemeColorLightest: '#fff2cc',
  hoverColorDark: '#e5ab00',
  hoverColorLight: '#ffc519',
  hoverColorLightest: '#ffebb2',
  textColorPrimary: '#4c3900',
  textColorSecondary: '#4c3900',
  textColorTertiary: '#7f5700',
  borderColor: '#cc9800',
}

export const PINK: ColorScheme = {
  schemeColorDarkest: '#e83f6f',
  schemeColorDark: '#ec658b',
  schemeColorLight: '#f18ba8',
  schemeColorLightest: '#fad8e2',
  hoverColorDark: '#d03863',
  hoverColorLight: '#ea527d',
  hoverColorLightest: '#f5b2c5',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#451221',
  borderColor: '#b93258',
}

export const BLUE: ColorScheme = {
  schemeColorDarkest: '#2274a5',
  schemeColorDark: '#4e8fb7',
  schemeColorLight: '#7aabc9',
  schemeColorLightest: '#d2e3ed',
  hoverColorDark: '#1e6894',
  hoverColorLight: '#3881ae',
  hoverColorLightest: '#bcd5e4',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#0d2e42',
  borderColor: '#1b5c84',
}

export const GREEN: ColorScheme = {
  schemeColorDarkest: '#00a323',
  schemeColorDark: '#32b54e',
  schemeColorLight: '#66c77b',
  schemeColorLightest: '#ccecd3',
  hoverColorDark: '#00921f',
  hoverColorLight: '#19ac38',
  hoverColorLightest: '#b2e3bd',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#00410e',
  borderColor: '#00821c',
}

export const AQUA: ColorScheme = {
  schemeColorDarkest: '#20e2e9',
  schemeColorDark: '#62eaef',
  schemeColorLight: '#8ff0f4',
  schemeColorLightest: '#d2f9fa',
  hoverColorDark: '#11cbd1',
  hoverColorLight: '#4ce7ed',
  hoverColorLightest: '#bcf6f8',
  textColorPrimary: '#094345',
  textColorSecondary: '#094345',
  textColorTertiary: '#094345',
  borderColor: '#19b4ba',
}

const colorSchemes: ColorScheme[] = [YELLOW, PINK, BLUE, GREEN, AQUA]

export default colorSchemes
