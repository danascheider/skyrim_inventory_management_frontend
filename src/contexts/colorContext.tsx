import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
