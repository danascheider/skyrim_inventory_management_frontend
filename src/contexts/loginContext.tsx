import { createContext, useState, useEffect, useCallback } from 'react'
import { type User } from 'firebase/auth'
import { useAuthState } from 'react-firebase-hooks/auth'
import { auth, signOutWithGoogle } from '../firebase'
import { ProviderProps } from '../types/contexts'

export interface LoginContextType {
  authLoading: boolean
  token: string | null
  withTokenRefresh: (fn: (idToken: string) => void) => void
  user?: User | null
}

export const LoginContext = createContext<LoginContextType>({
  user: null,
  token: null,
  authLoading: true,
  withTokenRefresh: () => {},
})

export const LoginProvider = ({ children }: ProviderProps) => {
  const [user, authLoading, authError] = useAuthState(auth)
  const [token, setToken] = useState<string | null>(null)

  const refreshToken = useCallback(() => {
    if (user) user.getIdToken(true).then((idToken) => setToken(idToken))
  }, [user])

  const withTokenRefresh = (fn: Function) => {
    if (user) {
      user.getIdToken(true).then((idToken) => {
        setToken(idToken)
        fn(idToken)
      })
    }
  }

  const value = {
    user,
    token,
    withTokenRefresh,
    authLoading,
  }

  useEffect(() => {
    if (authError) signOutWithGoogle()

    refreshToken()
  }, [user, authError])

  return <LoginContext.Provider value={value}>{children}</LoginContext.Provider>
}
