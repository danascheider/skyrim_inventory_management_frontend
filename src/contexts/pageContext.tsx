import { createContext, useState, useEffect, useRef } from 'react'
import { type ApiCalls, type Resource, type HttpVerb } from '../types/apiCalls'
import { ProviderProps } from '../types/contexts'
import { type FlashProps, type ModalProps } from '../types/pageContext'

interface PageContextType {
  flashProps: FlashProps
  setFlashProps: (props: FlashProps) => void
  modalProps: ModalProps
  setModalProps: (props: ModalProps) => void
  apiCallsInProgress: ApiCalls
  addApiCall: (key: Resource, value: HttpVerb) => void
  removeApiCall: (key: Resource, value: HttpVerb) => void
}

const defaultFlashProps: FlashProps = {
  type: 'info',
  message: '',
  hidden: true,
}

const defaultModalProps: ModalProps = {
  hidden: true,
  children: <></>,
}

const defaultApiCallStatus = {
  games: [],
  shoppingLists: [],
  shoppingListItems: [],
}

const PageContext = createContext<PageContextType>({
  flashProps: defaultFlashProps,
  modalProps: defaultModalProps,
  apiCallsInProgress: defaultApiCallStatus,
  setFlashProps: () => {},
  setModalProps: () => {},
  addApiCall: () => {},
  removeApiCall: () => {},
})

const PageProvider = ({ children }: ProviderProps) => {
  const [flashProps, setFlashProps] = useState<FlashProps>(defaultFlashProps)
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps)
  const [apiCallsInProgress, SetApiCallsInProgress] =
    useState<ApiCalls>(defaultApiCallStatus)

  const flashVisibleSince = useRef(0)

  const addApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: [...apiCallsInProgress[key], value],
    })
  }

  const removeApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: apiCallsInProgress[key].filter((verb) => verb !== value),
    })
  }

  const value = {
    flashProps,
    modalProps,
    apiCallsInProgress,
    setFlashProps,
    setModalProps,
    addApiCall,
    removeApiCall,
  }

  useEffect(() => {
    if (flashProps.hidden === true) return

    flashVisibleSince.current = Number(new Date())

    setTimeout(() => {
      if (Number(new Date()) - flashVisibleSince.current >= 4000) {
        setFlashProps({ ...flashProps, hidden: true })
        flashVisibleSince.current = 0
      }
    }, 4000)
  }, [flashProps])

  return <PageContext.Provider value={value}>{children}</PageContext.Provider>
}

export { PageContext, PageProvider }
