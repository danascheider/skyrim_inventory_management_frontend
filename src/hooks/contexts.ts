import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )