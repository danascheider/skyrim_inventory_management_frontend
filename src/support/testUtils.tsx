import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
