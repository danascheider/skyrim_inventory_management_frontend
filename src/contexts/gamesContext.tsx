import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
