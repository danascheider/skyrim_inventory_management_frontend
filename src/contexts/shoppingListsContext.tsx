import { createContext, useEffect, useState, useCallback } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type CallbackFunction } from '../types/functions'
import {
  type RequestShoppingListItem,
  type RequestShoppingList,
  type ResponseShoppingList,
} from '../types/apiData'
import { type ProviderProps } from '../types/contexts'
import { type ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import {
  postShoppingLists,
  getShoppingLists,
  patchShoppingList,
  deleteShoppingList,
  postShoppingListItems,
  deleteShoppingListItem,
} from '../utils/api/simApi'
import { useQueryString } from '../hooks/useQueryString'
import {
  useGoogleLogin,
  usePageContext,
  useGamesContext,
} from '../hooks/contexts'

const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface ShoppingListsContextType {
  shoppingLists: ResponseShoppingList[]
  shoppingListsLoadingState: LoadingState
  createShoppingList: (
    attributes: RequestShoppingList,
    onSuccess?: CallbackFunction,
    onError?: CallbackFunction
  ) => void
  updateShoppingList: (
    listId: number,
    attributes: RequestShoppingList,
    onSuccess?: CallbackFunction,
    onError?: CallbackFunction
  ) => void
  destroyShoppingList: (
    listId: number,
    onSuccess?: CallbackFunction,
    onError?: CallbackFunction
  ) => void
  createShoppingListItem: (
    listId: number,
    attributes: RequestShoppingListItem,
    onSuccess?: CallbackFunction,
    onError?: CallbackFunction
  ) => void
  destroyShoppingListItem: (
    itemId: number,
    onSuccess?: CallbackFunction,
    onError?: CallbackFunction
  ) => void
}

export const ShoppingListsContext = createContext<ShoppingListsContextType>({
  shoppingLists: [] as ResponseShoppingList[],
  shoppingListsLoadingState: LOADING,
  createShoppingList: () => {},
  updateShoppingList: () => {},
  destroyShoppingList: () => {},
  createShoppingListItem: () => {},
  destroyShoppingListItem: () => {},
})

export const ShoppingListsProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const { setFlashProps } = usePageContext()
  const { gamesLoadingState, games } = useGamesContext()
  const queryString = useQueryString()
  const [activeGame, setActiveGame] = useState<number | null>(null)
  const [shoppingListsLoadingState, setShoppingListsLoadingState] =
    useState(LOADING)
  const [shoppingLists, setShoppingLists] = useState<ResponseShoppingList[]>([])

  /**
   *
   * General handler for any ApiError
   *
   */

  const handleApiError = (e: ApiError, resource?: 'list' | 'list item') => {
    if (import.meta.env.DEV) console.error(e.message)

    if (e.code === 401) {
      signOutWithGoogle()
    } else if (e.code === 422) {
      setFlashProps({
        hidden: false,
        type: 'error',
        header: `${e.message.length} error(s) prevented your shopping ${
          resource || 'list'
        } from being saved:`,
        message: e.message,
      })
    } else {
      const message =
        e.code === 405
          ? "You can't manually manage an aggregate list."
          : UNEXPECTED_ERROR_MESSAGE

      setFlashProps({
        hidden: false,
        type: 'error',
        message,
      })
    }
  }

  /**
   *
   * Create shopping list for the active game
   *
   */

  const createShoppingList = useCallback(
    (
      attributes: RequestShoppingList,
      onSuccess?: CallbackFunction,
      onError?: CallbackFunction
    ) => {
      if (!activeGame) {
        setFlashProps({
          hidden: false,
          type: 'warning',
          message:
            'You must select a game from the dropdown before creating a shopping list.',
        })

        return
      }

      if (user && token) {
        postShoppingLists(activeGame, attributes, token)
          .then(({ json }) => {
            if (Array.isArray(json)) {
              if (json.length == 2) {
                setShoppingLists(json)
              } else {
                const newShoppingLists = shoppingLists
                newShoppingLists.splice(1, 0, json[0])
                setShoppingLists(newShoppingLists)
              }

              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your shopping list has been created.',
              })

              onSuccess && onSuccess()
            }
          })
          .catch((e: ApiError) => {
            if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The game you've selected doesn't exist, or doesn't belong to you. Please select another game and try again.",
              })
            } else {
              handleApiError(e)
            }

            onError && onError()
          })
      }
    },
    [user, token, activeGame, shoppingLists]
  )

  /**
   *
   * Fetch shopping lists for the active game
   *
   */

  const fetchShoppingLists = useCallback(() => {
    if (!activeGame) return

    if (user && token && activeGame) {
      setShoppingListsLoadingState(LOADING)

      getShoppingLists(activeGame, token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setShoppingLists(json)
            setShoppingListsLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (e.code === 404) {
            setFlashProps({
              hidden: false,
              type: 'error',
              message:
                "The game you've selected doesn't exist, or doesn't belong to you. Please select another game and try again.",
            })
          } else {
            handleApiError(e)
          }

          setShoppingLists([])
          setShoppingListsLoadingState(ERROR)
        })
    }
  }, [user, token, activeGame])

  /**
   *
   * Update specified shopping list
   *
   */

  const updateShoppingList = useCallback(
    (
      listId: number,
      attributes: RequestShoppingList,
      onSuccess?: CallbackFunction,
      onError?: CallbackFunction
    ) => {
      if (user && token) {
        patchShoppingList(listId, attributes, token)
          .then(({ status, json }) => {
            if (status === 200) {
              const newShoppingLists = [...shoppingLists]
              const index = newShoppingLists.findIndex(
                ({ id }) => id === listId
              )
              newShoppingLists[index] = json

              setShoppingLists(newShoppingLists)

              onSuccess && onSuccess()
            } else {
              // This won't happen but TypeScript doesn't know that
              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }
          })
          .catch((e: ApiError) => {
            if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The shopping list you tried to update doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e)
            }

            onError && onError()
          })
      }
    },
    [user, token, shoppingLists]
  )

  /**
   *
   * Destroy specified shopping list
   *
   */

  const destroyShoppingList = useCallback(
    (
      listId: number,
      onSuccess?: CallbackFunction,
      onError?: CallbackFunction
    ) => {
      if (user && token) {
        deleteShoppingList(listId, token)
          .then(({ json }) => {
            if ('errors' in json) {
              // This case should never happen because normally an ApiError
              // will be thrown for any response that includes this key, but
              // TypeScript doesn't know that.
              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            } else {
              const newShoppingLists = shoppingLists

              if (json.aggregate) newShoppingLists[0] = json.aggregate

              for (const deletedId of json.deleted) {
                const index = newShoppingLists.findIndex(
                  (list) => list?.id === deletedId
                )
                delete newShoppingLists[index]
              }

              setShoppingLists(newShoppingLists.filter((list) => !!list))
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your shopping list has been deleted.',
              })

              onSuccess && onSuccess()
            }
          })
          .catch((e: ApiError) => {
            if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The shopping list you tried to delete doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e)
            }

            onError && onError()
          })
      }
    },
    [user, token, shoppingLists]
  )

  /**
   *
   * Create a new shopping list item
   *
   */

  const createShoppingListItem = useCallback(
    (
      listId: number,
      attributes: RequestShoppingListItem,
      onSuccess?: CallbackFunction,
      onError?: CallbackFunction
    ) => {
      if (user && token) {
        postShoppingListItems(listId, attributes, token)
          .then(({ status, json }) => {
            if (status === 200 || status === 201) {
              const newShoppingLists = [...shoppingLists]

              for (let list of json) {
                const index = newShoppingLists.findIndex(
                  ({ id }) => id === list.id
                )
                newShoppingLists[index] = list
              }

              setShoppingLists(newShoppingLists)

              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your shopping list item has been created.',
              })

              onSuccess && onSuccess()
            } else {
              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }
          })
          .catch((e: ApiError) => {
            if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The shopping list you tried to add an item to doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e, 'list item')
            }

            onError && onError()
          })
      }
    },
    [user, token, shoppingLists]
  )

  /**
   *
   * Destroy a shopping list item
   *
   */

  const destroyShoppingListItem = useCallback(
    (
      itemId: number,
      onSuccess?: CallbackFunction,
      onError?: CallbackFunction
    ) => {
      if (user && token) {
        deleteShoppingListItem(itemId, token)
          .then(({ status, json }) => {
            if (status === 200) {
              const newShoppingLists = [...shoppingLists]

              newShoppingLists[0] = json[0]

              const index = newShoppingLists.findIndex(
                ({ id }) => id === json[1].id
              )
              newShoppingLists[index] = json[1]

              setShoppingLists(newShoppingLists)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your shopping list item was deleted.',
              })

              onSuccess && onSuccess()
            } else {
              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }
          })
          .catch((e: ApiError) => {
            if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "You have tried to delete a list item that doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e, 'list item')
            }

            onError && onError()
          })
      }
    },
    [user, token, shoppingLists]
  )

  /**
   *
   * Set the context provider value
   *
   */

  const value = {
    shoppingLists,
    shoppingListsLoadingState,
    createShoppingList,
    updateShoppingList,
    destroyShoppingList,
    createShoppingListItem,
    destroyShoppingListItem,
  }

  /**
   *
   * Set the active game automatically from the query string
   * or, failing that, from the games themselves when they load
   *
   */

  useEffect(() => {
    const gameId: number = Number(queryString.get('gameId'))

    if (gameId > 0) {
      setActiveGame(gameId)
    } else if (gamesLoadingState === DONE && games.length) {
      setActiveGame(games[0].id)
    }
  }, [queryString, gamesLoadingState, games])

  useEffect(() => {
    if (authLoading) return

    fetchShoppingLists()
  }, [authLoading, fetchShoppingLists])

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  return (
    <ShoppingListsContext.Provider value={value}>
      {children}
    </ShoppingListsContext.Provider>
  )
}
