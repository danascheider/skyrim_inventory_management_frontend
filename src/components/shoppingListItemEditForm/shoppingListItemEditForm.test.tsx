import { ReactElement } from 'react'
import { describe, expect, test, vitest } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { renderAuthenticated } from '../../support/testUtils'
import { BLUE } from '../../utils/colorSchemes'
import {
  gamesContextValue,
  shoppingListsContextValue,
} from '../../support/data/contextValues'
import { PageProvider } from '../../contexts/pageContext'
import { GamesContext } from '../../contexts/gamesContext'
import { ShoppingListsContext } from '../../contexts/shoppingListsContext'
import ShoppingListItemEditForm from './shoppingListItemEditForm'

let contextValue = shoppingListsContextValue

const renderWithContexts = (ui: ReactElement) =>
  renderAuthenticated(
    <PageProvider>
      <GamesContext.Provider value={gamesContextValue}>
        <ShoppingListsContext.Provider value={contextValue}>
          {ui}
        </ShoppingListsContext.Provider>
      </GamesContext.Provider>
    </PageProvider>
  )

describe('ShoppingListItemEditForm', () => {
  describe('displaying the form', () => {
    test('has the correct fields', () => {
      const wrapper = renderWithContexts(
        <ShoppingListItemEditForm
          itemId={6}
          description="Health potion ingredients"
          listTitle="Alchemy Ingredients"
          buttonColor={BLUE}
          quantity={2}
          unitWeight={3}
          notes={null}
        />
      )

      expect(wrapper.getByText('Health potion ingredients')).toBeTruthy()
      expect(wrapper.getByText('On list "Alchemy Ingredients"')).toBeTruthy()
      expect(wrapper.getByLabelText('Quantity').getAttribute('value')).toEqual(
        '2'
      )
      expect(
        wrapper.getByLabelText('Unit Weight').getAttribute('value')
      ).toEqual('3')
      expect(wrapper.getByLabelText('Notes')).toBeTruthy()
      expect(wrapper.getByText('Update Item')).toBeTruthy()
    })

    test('matches snapshot', () => {
      const wrapper = renderWithContexts(
        <ShoppingListItemEditForm
          itemId={6}
          description="Health potion ingredients"
          listTitle="Alchemy Ingredients"
          buttonColor={BLUE}
          quantity={2}
          unitWeight={3}
          notes={null}
        />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
