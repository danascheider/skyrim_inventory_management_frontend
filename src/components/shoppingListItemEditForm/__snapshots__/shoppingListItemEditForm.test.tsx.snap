// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingListItemEditForm > displaying the form > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_b0e6e0"
        style="--button-background-color: #2274a5; --button-text-color: #fff; --button-hover-color: #1e6894; --button-border-color: #1b5c84;"
      >
        <h3
          class="_header_b0e6e0"
        >
          Health potion ingredients
        </h3>
        <p
          class="_subheader_b0e6e0"
        >
          On list "Alchemy Ingredients"
        </p>
        <form
          data-testid="editShoppingListItem6Form"
        >
          <fieldset
            class="_fieldset_b0e6e0"
          >
            <label
              class="_label_b0e6e0"
            >
              Quantity
              <input
                class="_input_b0e6e0"
                inputmode="numeric"
                min="1"
                pattern="^[0-9]*$"
                placeholder="Quantity"
                required=""
                title="Quantity must be an integer greater than 0"
                type="number"
                value="2"
              />
            </label>
          </fieldset>
          <fieldset
            class="_fieldset_b0e6e0"
          >
            <label
              class="_label_b0e6e0"
            >
              Unit Weight
              <input
                class="_input_b0e6e0"
                inputmode="numeric"
                min="0"
                placeholder="Unit Weight"
                step="0.1"
                type="number"
                value="3"
              />
            </label>
          </fieldset>
          <fieldset
            class="_fieldset_b0e6e0"
          >
            <label
              class="_label_b0e6e0"
            >
              Notes
              <input
                class="_input_b0e6e0"
                placeholder="Notes"
                type="text"
                value=""
              />
            </label>
          </fieldset>
          <button
            class="_submit_b0e6e0"
            type="submit"
          >
            Update Item
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_b0e6e0"
      style="--button-background-color: #2274a5; --button-text-color: #fff; --button-hover-color: #1e6894; --button-border-color: #1b5c84;"
    >
      <h3
        class="_header_b0e6e0"
      >
        Health potion ingredients
      </h3>
      <p
        class="_subheader_b0e6e0"
      >
        On list "Alchemy Ingredients"
      </p>
      <form
        data-testid="editShoppingListItem6Form"
      >
        <fieldset
          class="_fieldset_b0e6e0"
        >
          <label
            class="_label_b0e6e0"
          >
            Quantity
            <input
              class="_input_b0e6e0"
              inputmode="numeric"
              min="1"
              pattern="^[0-9]*$"
              placeholder="Quantity"
              required=""
              title="Quantity must be an integer greater than 0"
              type="number"
              value="2"
            />
          </label>
        </fieldset>
        <fieldset
          class="_fieldset_b0e6e0"
        >
          <label
            class="_label_b0e6e0"
          >
            Unit Weight
            <input
              class="_input_b0e6e0"
              inputmode="numeric"
              min="0"
              placeholder="Unit Weight"
              step="0.1"
              type="number"
              value="3"
            />
          </label>
        </fieldset>
        <fieldset
          class="_fieldset_b0e6e0"
        >
          <label
            class="_label_b0e6e0"
          >
            Notes
            <input
              class="_input_b0e6e0"
              placeholder="Notes"
              type="text"
              value=""
            />
          </label>
        </fieldset>
        <button
          class="_submit_b0e6e0"
          type="submit"
        >
          Update Item
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
